In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [ ]:
df = pd.read_csv('/content/IOT_Devices_data.xlsx - Sheet1.csv')
df.head()

id         Device ID            Timestamp  Temperature (C)  \
0   1  Smart Thermostat  2023-01-01 23:00:00            23.28   
1   1  Smart Thermostat  2023-01-02 00:00:00            19.59   
2   1  Smart Thermostat  2023-01-02 23:00:00            23.00   
3   1  Smart Thermostat  2023-01-04 11:00:00            20.30   
4   1  Smart Thermostat  2023-01-04 13:00:00            33.05   

   Power Consumption (W) Error Code Maintenance Required  
0                 118.49        NaN                   No  
1                 106.04        NaN                   No  
2                 104.11        NaN                   No  
3                 100.27        NaN                   No  
4                 134.59        NaN                   No

In [ ]:
df['Maintenance Required'].value_counts()

Maintenance Required
No     1426
Yes      74
Name: count, dtype: int64

In [ ]:
df_maintenance_required = df[df['Maintenance Required'] == "Yes"]

In [ ]:
df_new = df[df['Maintenance Required'] == "No"][:100]

In [ ]:
df_maintenance_required.shape

(74, 7)

In [ ]:
df.head()

id         Device ID            Timestamp  Temperature (C)  \
0   1  Smart Thermostat  2023-01-01 23:00:00            23.28   
1   1  Smart Thermostat  2023-01-02 00:00:00            19.59   
2   1  Smart Thermostat  2023-01-02 23:00:00            23.00   
3   1  Smart Thermostat  2023-01-04 11:00:00            20.30   
4   1  Smart Thermostat  2023-01-04 13:00:00            33.05   

   Power Consumption (W) Error Code Maintenance Required  
0                 118.49        NaN                   No  
1                 106.04        NaN                   No  
2                 104.11        NaN                   No  
3                 100.27        NaN                   No  
4                 134.59        NaN                   No

In [ ]:
df_new = pd.concat([df_maintenance_required, df_new])

In [ ]:
df_new.head()

id         Device ID            Timestamp  Temperature (C)  \
26    1  Smart Thermostat  2023-01-17 23:00:00            31.27   
42    1  Smart Thermostat  2023-01-29 16:00:00            32.61   
48    1  Smart Thermostat  2023-02-02 16:00:00            22.09   
65    1  Smart Thermostat  2023-02-12 04:00:00            28.56   
123   2   Security Camera  2023-01-22 11:00:00            28.85   

     Power Consumption (W) Error Code Maintenance Required  
26                  109.34     Error5                  Yes  
42                   95.36     Error5                  Yes  
48                  118.62     Error2                  Yes  
65                  101.93     Error4                  Yes  
123                 124.77     Error4                  Yes

In [ ]:
df_new['Maintenance Required'].value_counts()

Maintenance Required
No     100
Yes     74
Name: count, dtype: int64

In [ ]:
df_new['Maintenance Required'] = le.fit_transform(df_new['Maintenance Required'])

In [ ]:
df_new_1 = pd.get_dummies(df_new.drop(['Timestamp','Maintenance Required'],axis = 1))

In [ ]:
df_new_1.head()

id  Temperature (C)  Power Consumption (W)  Device ID_Garage Door Opener  \
26    1            31.27                 109.34                         False   
42    1            32.61                  95.36                         False   
48    1            22.09                 118.62                         False   
65    1            28.56                 101.93                         False   
123   2            28.85                 124.77                         False   

     Device ID_Motion Sensor  Device ID_Security Camera  \
26                     False                      False   
42                     False                      False   
48                     False                      False   
65                     False                      False   
123                    False                       True   

     Device ID_Smart Door Lock  Device ID_Smart Lighting System  \
26                       False                            False   
42                       False                            False   
48                       False                            False   
65                       False                            False   
123                      False                            False   

     Device ID_Smart Oven  Device ID_Smart Plug  ...  \
26                  False                 False  ...   
42                  False                 False  ...   
48                  False                 False  ...   
65                  False                 False  ...   
123                 False                 False  ...   

     Device ID_Smart Thermostat  Device ID_Smoke Detector  \
26                         True                     False   
42                         True                     False   
48                         True                     False   
65                         True                     False   
123                       False                     False   

     Device ID_Video Doorbell  Device ID_Water Leak Detector  \
26                      False                          False   
42                      False                          False   
48                      False                          False   
65                      False                          False   
123                     False                          False   

     Device ID_Window/Door Sensor  Error Code_Error1  Error Code_Error2  \
26                          False              False              False   
42                          False              False              False   
48                          False              False               True   
65                          False              False              False   
123                         False              False              False   

     Error Code_Error3  Error Code_Error4  Error Code_Error5  
26               False              False               True  
42               False              False               True  
48               False              False              False  
65               False               True              False  
123              False               True              False  

[5 rows x 23 columns]

In [ ]:
df_new_1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 174 entries, 26 to 103
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   id                               174 non-null    int64  
 1   Temperature (C)                  174 non-null    float64
 2   Power Consumption (W)            174 non-null    float64
 3   Device ID_Garage Door Opener     174 non-null    bool   
 4   Device ID_Motion Sensor          174 non-null    bool   
 5   Device ID_Security Camera        174 non-null    bool   
 6   Device ID_Smart Door Lock        174 non-null    bool   
 7   Device ID_Smart Lighting System  174 non-null    bool   
 8   Device ID_Smart Oven             174 non-null    bool   
 9   Device ID_Smart Plug             174 non-null    bool   
 10  Device ID_Smart Refrigerator     174 non-null    bool   
 11  Device ID_Smart Speaker          174 non-null    bool   
 12  Device ID_Smart TV        

In [ ]:

import numpy as np
df_new_1 = df_new_1.astype(np.int64)
df_new_1.head()

id  Temperature (C)  Power Consumption (W)  Device ID_Garage Door Opener  \
26    1               31                    109                             0   
42    1               32                     95                             0   
48    1               22                    118                             0   
65    1               28                    101                             0   
123   2               28                    124                             0   

     Device ID_Motion Sensor  Device ID_Security Camera  \
26                         0                          0   
42                         0                          0   
48                         0                          0   
65                         0                          0   
123                        0                          1   

     Device ID_Smart Door Lock  Device ID_Smart Lighting System  \
26                           0                                0   
42                           0                                0   
48                           0                                0   
65                           0                                0   
123                          0                                0   

     Device ID_Smart Oven  Device ID_Smart Plug  ...  \
26                      0                     0  ...   
42                      0                     0  ...   
48                      0                     0  ...   
65                      0                     0  ...   
123                     0                     0  ...   

     Device ID_Smart Thermostat  Device ID_Smoke Detector  \
26                            1                         0   
42                            1                         0   
48                            1                         0   
65                            1                         0   
123                           0                         0   

     Device ID_Video Doorbell  Device ID_Water Leak Detector  \
26                          0                              0   
42                          0                              0   
48                          0                              0   
65                          0                              0   
123                         0                              0   

     Device ID_Window/Door Sensor  Error Code_Error1  Error Code_Error2  \
26                              0                  0                  0   
42                              0                  0                  0   
48                              0                  0                  1   
65                              0                  0                  0   
123                             0                  0                  0   

     Error Code_Error3  Error Code_Error4  Error Code_Error5  
26                   0                  0                  1  
42                   0                  0                  1  
48                   0                  0                  0  
65                   0                  1                  0  
123                  0                  1                  0  

[5 rows x 23 columns]

In [ ]:
X = df_new_1.drop(['id'],axis=1)
Y = df_new['Maintenance Required']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

In [ ]:
models_list = [LogisticRegression(), SVC(), RandomForestClassifier(), KNeighborsClassifier()]

In [ ]:
for model in models_list:
  model.fit(X_train, y_train)
  y_pred_train = model.predict(X_train)
  score = accuracy_score(y_train, y_pred_train)

  print(f"{model} got {score} accuracy")

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression() got 0.9640287769784173 accuracy
SVC() got 0.60431654676259 accuracy
RandomForestClassifier() got 1.0 accuracy
KNeighborsClassifier() got 0.7410071942446043 accuracy


LogisticRegression() got 0.9640287769784173 accuracy
SVC() got 0.60431654676259 accuracy
RandomForestClassifier() got 1.0 accuracy
KNeighborsClassifier() got 0.7410071942446043 accuracy


so with high train accuracy RandomForestClassifier won

In [ ]:
model_2 = RandomForestClassifier()
model_2.fit(X_train, y_train)
Y_pred_test = model_2.predict(X_test)
score = accuracy_score(y_test, Y_pred_test)
print(score)

1.0
